src/orange/BoundingBox.hh

```diff
@@ -101,40 +142,55 @@ template<class T>
 CELER_FUNCTION BoundingBox<T>::BoundingBox(Real3 const& lo, Real3 const& hi)
     : lower_(lo), upper_(hi)
 {
-    CELER_EXPECT(lower_[0] <= upper_[0] && lower_[1] <= upper_[1]
-                 && lower_[2] <= upper_[2]);
+    if constexpr (CELERITAS_DEBUG)
+    {
+        for (auto ax : {Axis::x, Axis::y, Axis::z})
+        {
+            CELER_EXPECT(lower_[to_int(ax)] <= upper_[to_int(ax)]);
+        }
+    }
+    CELER_ENSURE(*this);
 }
 
 //---------------------------------------------------------------------------//
 /*!
- * Lower bbox coordinate (must be valid).
+ * Create a possibly degenerate bounding box from two points.
  */
 template<class T>
-CELER_FUNCTION auto BoundingBox<T>::lower() const -> Real3 const&
+CELER_CONSTEXPR_FUNCTION
+BoundingBox<T>::BoundingBox(std::true_type, Real3 const& lo, Real3 const& hi)
+    : lower_(lo), upper_(hi)
 {
-    CELER_EXPECT(*this);
-    return lower_;
 }
 
 //---------------------------------------------------------------------------//
 /*!
- * Upper bbox coordinate (must be valid).
+ * Whether the bbox is in a nondegenerate state.
  */
 template<class T>
-CELER_FUNCTION auto BoundingBox<T>::upper() const -> Real3 const&
+CELER_CONSTEXPR_FUNCTION BoundingBox<T>::operator bool() const
 {
-    CELER_EXPECT(*this);
-    return upper_;
+    return lower_[to_int(Axis::x)] <= upper_[to_int(Axis::x)]
```

Comment:
These need to be `<` not `<=` to differentiate between null and degenerate bboxes. Consider the bbox: `{{1, 1, 1}, {2, 2, 2}}`. If you intersect it a null bbox, it should be a no-op. However, if you intersect it with `{{0,0,0}, {0, 0, 0}}`, you should get `{{0, 0, 0}, {2, 2, 2}}`. I could see problems arise downstream from not differentiating between these two cases due to bool returning false in both cases.

Reply:
If you *intersect* you should get a degenerate bbox, because no point is in both 1 and 2. It can't be simply degenerate because those bounding boxes do contain points that are coincident on the faces.

If you *union* you get `{{0, 0, 0}, {2, 2, 2}}`, which is perhaps what you mean?

I'm adding more tests, comments, and examples to demonstrate all this.